pkg/api/pod/util.go

```diff
@@ -386,6 +386,7 @@ func GetValidationOptionsFromPodSpecAndMeta(podSpec, oldPodSpec *api.PodSpec, po
 		AllowNamespacedSysctlsForHostNetAndHostIPC:        false,
 		AllowNonLocalProjectedTokenPath:                   false,
 		AllowImageVolumeSource:                            utilfeature.DefaultFeatureGate.Enabled(features.ImageVolume),
+		AllowPodLifecycleSleepActionZeroValue:             utilfeature.DefaultFeatureGate.Enabled(features.PodLifecycleSleepActionAllowZero),
```

Comment:
The usual pattern is to initialize this, then (below) do something like:

```
opts.AllowPodLifecycleSleepActionZeroValue = opts.AllowPodLifecycleSleepActionZeroValue || lifecycleSleepZeroUsed(podSpec)
```

Where:

```
func lifecycleSleepZeroUsed(spec *api.PodSpec) bool {
    for i := range spec.containers {
        c := &spec.Containers[i]
        if c.Lifecycle != nil {
            if c.Lifecycle.PostStart != nil && c.Lifecycle.PostStart.Sleep != nil && c.Lifecycle.PostStart.Sleep.Seconds == 0 {
                return true
            }
            if c.Lifecycle.PreStop != nil && c.Lifecycle.PreStop.Sleep != nil && c.Lifecycle.PreStop.Sleep.Seconds == 0 {
                return true
            }
        }
    }
    return false
}
```

There's a util function somewhere, I can't recall the name, that lets you easily repeat this for spec.Containers, .EphemeralContainers, and .InitContainers.  You should use that.

This allows us to respect existing pods which have set this value.  See https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/feature-gates.md for more

Reply:
I believe the util function is [visitContainers](https://github.com/kubernetes/kubernetes/blob/master/pkg/api/pod/util.go#L63).
https://github.com/kubernetes/kubernetes/blob/f153edf356324564396ae3f20538859783aadc76/pkg/api/pod/util.go#L63

It makes sense how by setting the following line `opts.AllowPodLifecycleSleepActionZeroValue = opts.AllowPodLifecycleSleepActionZeroValue || lifecycleSleepZeroUsed(podSpec)`, we can respect existing Pods with the value set to zero and make updates work and tolerate the zero value like you commented in the KEP Google doc. 

I can see a function with a similar pattern for the parent KEP too. This is really cool!
https://github.com/kubernetes/kubernetes/blob/f153edf356324564396ae3f20538859783aadc76/pkg/api/pod/util.go#L776-L781

I made the changes to add the `lifecycleSleepZeroUsed` function. I'm trying to test it locally but kind is throwing some errors. Let me take a look. I will push the code changes here as well as the KEP PR with the changes soon.